Specified Monin-Obukhov Length

hgopalan opened the pull request:
There are three ways to specify non-neutral stratification for ABL simulations: (i) specified heat-flux, (ii) specified temperature or heating/cooling rate and (iii) specified Monin-Obukhov length. This PR adds the capability for the third methodology in which the Monin-Obukhov length is kept constant. This method is used often in RANS simulations to do wind turbine siting for non-neutral cases (faster convergence rate) and minimize the mismatch between wall boundary condition and inflow profile. 

This method is also robust for running SBL cases in which the non-linear iteration for MOL may not converge. 


Please check the type of change introduced:

- [ ] Bugfix
- [X ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Checklist

The following is included:

- [ ] new unit-test(s)
- [x] new regression test(s)
- [ ] documentation for new capability

This PR was tested by running:

- the unit tests
  - [ ] on GPU <!-- note the OS and compiler -->
  - [X ] on CPU <!-- note the OS and compiler -->
- the regression tests
  - [ ] on GPU <!-- note the OS and compiler -->
  - [ ] on CPU <!-- note the OS and compiler -->

hgopalan commented:
Need additional test cases for immersed forcing

moprak-nrel commented on amr-wind/equation_systems/temperature/source_terms/DragTempForcing.H:
```diff
@@ -33,6 +34,50 @@ private:
+    amrex::Real m_gamma_h{5.0};
+    amrex::Real m_beta_h{16.0};
+
+    static amrex::Real stability(
```
This function seems to be getting repeated implementations across multiple files? Just have one and call those?

hgopalan replied:
I did not know how to do that across namespaces so created one at every place.

moprak-nrel replied:
You could just put it in a common place and call it across the places you need. For instance you could add these in some type of wall function utilities. We don't want to be repeating code.

moprak-nrel commented on test/test_files/abl_specifiedmol_neutral/rans_1d.info:
Is the rans_1d.info file identical across the cases? You might be able to just have one and reuse it for the other tests?

hgopalan replied:
No they are different for each case.

moprak-nrel commented on amr-wind/equation_systems/icns/source_terms/DragForcing.cpp:
```diff
@@ -181,9 +192,16 @@ void DragForcing::operator()(
-    const auto tiny = std::numeric_limits<amrex::Real>::epsilon();
     const amrex::Real kappa = 0.41;
     const amrex::Real cd_max = 1000.0;
+    amrex::Real non_neutral_neighbour = 0.0;
```
Can you const these with a ternary expression?

moprak-nrel commented on amr-wind/equation_systems/icns/source_terms/DragForcing.cpp:
```diff
@@ -251,17 +269,20 @@ void DragForcing::operator()(
-                uTarget * ux2r / (tiny + std::sqrt(ux2r * ux2r + uy2r * uy2r));
-            const amrex::Real uyTarget =
-                uTarget * uy2r / (tiny + std::sqrt(ux2r * ux2r + uy2r * uy2r));
+                ustar / kappa * (std::log(0.5 * dx[2] / z0) - non_neutral_cell);
```
Are these calcs/the model going to be in the documentation? Not sure how to interpret `non_neutral_cell` and `non_neutral_neighbor`.

hgopalan replied:
The walk through talks about the details already. The non_neutral are stability function corrections.

moprak-nrel commented on amr-wind/equation_systems/temperature/source_terms/DragTempForcing.cpp:
```diff
@@ -46,24 +56,58 @@ void DragTempForcing::operator()(
     amrex::FArrayBox ref_theta_fab(bx, 1, amrex::The_Async_Arena());
     amrex::Array4<amrex::Real> const& ref_theta_arr = ref_theta_fab.array();
     m_transport.ref_theta_impl(lev, mfi, bx, ref_theta_arr);
+    const amrex::Real gravity_mod = 9.81;
```
Do you want to be using the value of gravity from the input file? We have `incflo.gravity` or some such right?

moprak-nrel commented on amr-wind/equation_systems/tke/source_terms/KransAxell.H:
```diff
@@ -65,6 +65,28 @@ private:
+    amrex::Real m_gamma_m{5.0};
+    amrex::Real m_beta_m{16.0};
+
+    static amrex::Real stability(
```
Same comment as repeated definition of this function

moprak-nrel commented on amr-wind/wind_energy/ABLWallFunction.cpp:
```diff
@@ -331,20 +389,61 @@ void ABLTempWallFunc::wall_model(
+                    if (zeta > 0) {
+                        psi_m = -m_gamma_m * zeta;
+                    } else {
+                        amrex::Real x =
```
This variable name is confusing, this is not `plo + i*dx`.

hgopalan replied:
Copied from here: https://github.com/Exawind/amr-wind/blob/main/amr-wind/wind_energy/MOData.cpp

moprak-nrel replied:
Why not just call that function?

moprak-nrel replied:
Also context matters, that is a standalone small function where `x` isn't going to get confused with the spatial direction. Which isn't the case in your code.

hgopalan replied:
> Why not just call that function?

I want to have the flexibility of using a different stability function in the future within the context of this model. Currently it is set to the same as the one in MODData.H.

moprak-nrel replied:
In that case, create a function for calculating `psi_m/psi_h` in your namespace and call the `MOData` function within that. Or simply just call the MOData function for now, and replace it when you add a different stability function.

moprak-nrel commented on amr-wind/wind_energy/ABLWallFunction.cpp:
```diff
@@ -331,20 +389,61 @@ void ABLTempWallFunc::wall_model(
+                            const amrex::Real vv = vold_arr(i, j, k, 1);
+                            const amrex::Real wspd =
+                                std::sqrt(uu * uu + vv * vv);
+                            const amrex::Real theta2 = told_arr(i, j, k);
```
maybe a more descriptive variable name? you have `theta2` to refer to `theta` at k+1 in a different place.

hgopalan replied:
In this case the value at k is correct so I am referring to it as theta2.

hgopalan commented on amr-wind/equation_systems/temperature/source_terms/DragTempForcing.cpp:
```diff
@@ -46,24 +56,58 @@ void DragTempForcing::operator()(
+        //! We do not know the actual temperature so use cell above
+        const amrex::Real thetastar =
+            theta * ustar * ustar / (kappa * gravity_mod * mol_length);
+        const amrex::Real surf_temp =
```
For terrain, the value at the point above the cell is computed from the solution as it is not a wall. However, we need to force a MOST boundary condition as a forcing term. So we correct the theta at the cell using the cell above it by computing what is the MOST solution. We already had this in place for wind speed. It gets more complicated with temperature.

marchdf commented on docs/sphinx/user/inputs_ABL.rst:
```diff
@@ -220,3 +220,18 @@ This section is for setting atmospheric boundary layer parameters.
 
    `ICNS.source_terms  = VelocityFreeAtmosphereForcing`
 
+.. input_param:: ABL.wall_het_model
```
what is `het` ? Can we make it a more explicit user param?

hgopalan replied:
het to represent heterogenous model. I can expand it out.

marchdf commented on docs/sphinx/user/inputs_ABL.rst:
```diff
@@ -220,3 +220,18 @@ This section is for setting atmospheric boundary layer parameters.
+   Allows the use of different surface model options for the Monin-Obukhov length. Currently supports two options:
+   (i) "none" - original model in the code and (ii) "mol" - Monin-Obukhov length is constant while heat-flux varies 
+
+.. input_param:: ABL.mol_length
```
what is `mol`? Worried this user facing param is going to conflict with the MOL advection numerics option.

hgopalan replied:
mol for Monin-Obuknov length. We can change it to a different name if necessary

marchdf replied:
that would be great!

marchdf commented on amr-wind/wind_energy/ABLWallFunction.H:
```diff
@@ -98,6 +104,14 @@ public:
     const ABLWallFunction& m_wall_func;
     std::string m_wall_shear_stress_type{"moeng"};
+    std::string m_wall_het_model{"none"};
+    amrex::Real m_mol_length{-1e30};
```
instead of `-1e30` can you use a value from the `constants` namespace?

moprak-nrel commented:
@hgopalan I've done a bunch of the C++ lift, and addressed comments from me and @marchdf. Can you take a look to make sure I haven't made any behavior changes?

Also can you address the merge conflicts? Seems like there's a conflict in `DragForcing` that needs to be addressed. 

We also need @marchdf  or someone else to review since I've become a significant contributor to this PR ;) 

Final minor point, the regtest `abl_specifiedmol_neutral` does not have a `monin_obuknov_length` specified in the input, is this intended? That's the only diff I see before and after my changes in the new regtests.

hgopalan commented:
> @hgopalan I've done a bunch of the C++ lift, and addressed comments from me and @marchdf. Can you take a look to make sure I haven't made any behavior changes?
> 
> Also can you address the merge conflicts? Seems like there's a conflict in `DragForcing` that needs to be addressed.
> 
> We also need @marchdf or someone else to review since I've become a significant contributor to this PR ;)
> 
> Final minor point, the regtest `abl_specifiedmol_neutral` does not have a `monin_obuknov_length` specified in the input, is this intended? That's the only diff I see before and after my changes in the new regtests.

For neutral cases we do not need it.

mbkuhn commented on amr-wind/equation_systems/icns/source_terms/DragForcing.H:
```diff
@@ -50,8 +51,15 @@ private:
     int m_sponge_south{0};
     int m_sponge_north{1};
     bool m_is_laminar{false};
+    
```
```suggestion
```

mbkuhn commented on amr-wind/equation_systems/icns/source_terms/DragForcing.H:
```diff
@@ -50,8 +51,15 @@ private:
+    
     bool m_terrain_is_waves{false};
     bool m_apply_MOSD{false};
+    
```
```suggestion
```
